Add option to open URLs on click

chrisduerr opened the pull request:
This adds the option to automatically launch URLs with a specified
program when clicking on them.

The config option `mouse.url_launcher` has been added to specify which
program should be used to open the URL. The URL is always passed as the
last parameter to the specified command.

This fixes #113.

chrisduerr commented on alacritty.yml:
```diff
@@ -259,6 +259,15 @@ mouse:
+  # This program is executed when clicking on a text which is recognized as a URL.
+  # The URL is always added to the command as the last parameter.
+  #url_launcher:
+  #  program: /usr/bin/firefox
```
Yeah, I've thought about that and it would probably be a good idea. I just don't have it installed to test it, but I'll give it a go to see which parameters might be required.

nixpulvis commented:
This is pretty exciting. Just a few comments on the UX.

- The cursor should turn into a pointer when over a URL, to indicate to a user that they can click it.
- It would be really cool if the URL was underlined on hover, to help users see what the link is parsed as.

I've seen this functionality tied to a modifier key. Would this be something we want to add as well. I think I personally would want to put this behind <kbd>shift</kbd> or <kbd>ctrl</kbd>.

nixpulvis commented:
Also, how do we feel about parsing paths as `file://<path>`?

nixpulvis commented:
Please ignore my comments that are covered by #1618, I didn't see that until now.

chrisduerr commented:
@nixpulvis `file:///home/user/stuff` should already be recognized since it's a valid URL. The only question is if applications like `xdg-open` would open such a URL with a browser or the local file manager.

nixpulvis commented:
@chrisduerr I should have given an example. I'm talking about opening strings like `/home/nixpulvis/todo.txt`. As far as the question of what program should be opened. This is all configurable with `xdg-open`, less sure about macOS.

Currently `xdg-open` will open URLs in Firefox, and files in `vim`.

nixpulvis commented:
I agree this is a separate feature, and I'm not sure of other terminals that do this. Just FWIW, I was imagining it would validate the paths and just call `xdg-open` (or `open` on macOS) like anything else. For me this would open them in `vim` as opposed to a file manager, which would be pretty neat.

nixpulvis commented:
There's currently a bug with the indexing and scrollback. If the URL was scrolled back to at all the iterators start from the wrong place.

Another potential issue is with `URL_SEPARATOR_CHARS`. Detecting URLs is tricky stuff, I won't pretend to be an expert in all the edge cases, but there's a decent blog post about it by Jeff Atwood:https://blog.codinghorror.com/the-problem-with-urls/. One example we currently fail is `echo "(https://nixpulvis.com)"`.

nixpulvis commented:
Also, I'm not sure I'd want URL launching to happen inside `vim` by simply clicking (left or right ATM) on a URL. I'm not sure if there's a good way to handle this, other than to support modifier keys (/ right click?) from the start.

kpcyrd commented:
@nixpulvis other terminal emulators do this with `ctrl`+`left-click`

chrisduerr commented:
I've looked into other terminal emulators and it seems like VTE (Termite) just allows all parantheses inside the URL, so `(https://github.com)` will incorrectly open `https://github.com)`.

However URxvt seems to be a bit more advanced in this case. URxvt seems to allow parentheses under the condition that they are matched. So `(https://github.com)` is correctly recognized, but so is `https://en.wikipedia.org/wiki/Rust_(video_game)`.

I think it should be easily possible to mimic this behavior by breaking on `)`, but only if there was no matching `(` previously. I'll try to play with it and add it to this PR.

kpcyrd commented:
@chrisduerr nice! I think we should have a few tests, eg for the urls you mentioned to ensure this doesn't break.

I'm not familiar with alacritty internals, but:
```
const URL_SEPARATOR_CHARS: [char; 3] = [' ', '"', '\''];
```
wouldn't this match tabs or non-printable characters like terminal control characters?

csmoe commented on src/config.rs:
```diff
@@ -744,6 +749,22 @@ enum CommandWrapper {
 }
 
+impl CommandWrapper {
+    pub fn program(&self) -> &String {
```
`&String`: why not `&str`?

csmoe commented on src/config.rs:
```diff
@@ -744,6 +749,22 @@ enum CommandWrapper {
+    pub fn program(&self) -> &String {
+        match *self {
+            CommandWrapper::Just(ref program) => program,
+            CommandWrapper::WithArgs { ref program, .. } => program,
```
you deref self and then `ref` it again, it seems not needed, just match on self directly, you'll get the reference:
```rust
match self {
     CommandWrapper::Just(program) => program,
}
```
(btw, `ref` is not "recommended" anymore since [1.26](https://github.com/rust-lang/rust/pull/49394))

chrisduerr replied:
Old habits die hard :D. Thanks for the review!

chrisduerr commented:
The feedback from @csmoe has been incorporated into the last commit.

I've also added some tests to make sure the basic functionality is protected against regression.

Since it has been asked for by @nixpulvis, it is now also possible to specify modifier keys which need to be held down when clicking on URLs. The config now has the fields `mouse.url.launcher` for the application which should launch the URL and `mouse.url.modifiers` for the modifiers which need to be held.

This should resolve all the feedback I've seen on this feature so far. If you've got any other ideas or complaints, please let me know.

NickeZ commented on src/term/mod.rs:
```diff
@@ -80,6 +96,52 @@ impl selection::SemanticSearch for Term {
+        // Heuristic to remove all leading '('
+        while buf.starts_with('(') {
+            buf.remove(0);
+        }
```
This seems quite inefficient to me. `remove(0)` is `O(n)`. Wouldn't it make sense to use `trim_left_matches()`?

chrisduerr replied:
I wasn't much concerned with efficiency here, since it's only executed on click. But if this is re-used from a different code, that seems like a worthy cause.

I've looked at the `trim_left_matches` function, since I wasn't aware that exists, and gave it a quick spin just to confirm that it's faster. However I noticed that it is not.

If we're just comparing str trimming to str trimming, the `trim_left_matches` function is faster. However, we're trimming a `String`, not a `&str`. And the function returns a `&'a str`. So to make it work we'd have to change it to `trim_left_matches('(').into()`, which is super inefficient because of the `into`.

Here's the code I've used if you want to benchmark it yourself:
```
#![feature(test)]

extern crate test;

#[bench]
fn tlm(b: &mut test::Bencher) {
    b.iter(|| {
        let mut test = String::from("((Hello, World!");
        test = test.trim_left_matches('(').to_owned();
        test::black_box(test);
    });
}

#[bench]
fn other(b: &mut test::Bencher) {
    b.iter(|| {
        let mut test = String::from("((Hello, World!");
        while test.starts_with('(') {
            test.remove(0);
        }
        test::black_box(test);
    });
}
```

And these are the results I've got:
```
running 2 tests
test other ... bench:          26 ns/iter (+/- 0)
test tlm   ... bench:          42 ns/iter (+/- 0)
```

Note that I'm not saying that `remove(0)` is the most efficient way to do it here. I think it should be fine since we're usually trimming 0 characters, a realistic worse case might be 2-3. So checking first if any trimming is required at all probably resolves most performance issues.

If you have any other idea how to make this fast **and** easy to read, let me know. For now I think this should be an appropriate representation of what we're trying to achieve.

NickeZ replied:
Thanks for the exhaustive answer! I didn't consider the fact that it was an owned string and that the trimming function was defined for `str`. I also thought that it was more expensive for `remove(0)` to move the whole string.

chrisduerr commented on CHANGELOG.md:
```diff
@@ -19,6 +19,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Add `save_to_clipboard` configuration option for copying selected text to the system clipboard
 - New terminfo entry, `alacritty-direct`, that advertises 24-bit color support
 - Add support for CSI sequences Cursor Next Line (`\e[nE`) and Cursor Previous Line (`\e[nF`)
+- When `mouse.url_launcher` is set, clicking on URLs will now open them with the specified program
```
Thanks for catching this.

chrisduerr commented:
Waiting for #1374 before revisiting this since it's a major change. I think as an initial implementation this should go in the right direction though.

zacps commented:
The launcher on windows should be `start`, I don't think there's any other platform specific code needed.

zacps commented:
The launcher on windows should be `explorer`.

chrisduerr commented on src/event.rs:
```diff
@@ -196,6 +200,7 @@ pub struct Mouse {
     pub column: Column,
     pub cell_side: Side,
     pub lines_scrolled: f32,
+    pub last_press_pos: (usize, usize),
```
I don't like the idea of using `last_press_pos` here to check that the URL has been clicked rather than dragged.

Though I'm not sure if there's any better solution for this problem.

chrisduerr replied:
Maybe using a boolean to indicate that the mouse has moved since the last button down is a bit nicer. This could be updated in the mouse_moved method.

Checking that the cursor moved at least on cell seems reasonable. Also it would still allow single-cell selection to be recognized as a URL click.

chrisduerr commented:
I've refactored the code a little bit and added the configuration option to Windows so everything should work nicely now.

As far as I'm concerned this should be good to go unless any more tests fail in CI. If someone notices any issues during testing, please let me know.

jwilm commented:
@kpcyrd thanks for the feedback! Those notes are great, and it would be super helpful if you could file an issue for URL handling improvements so that we can better track a resolution.

chrisduerr commented:
> Links aren't underlined when hovering over them. This is more of a feature request, but this is important because it shows you what it's actually going to pass to xdg-open

See #1618.

> I think mouse.url.modifiers should be set by default to avoid interfering with a semantic click

This depends on how you're using this feature I guess. I personally don't want it behind any modifier which is why I've made that the default, however your argument about semantic selection is very valid.

I always like to imitate the behavior of my previous terminal emulator Termite (VTE), which behaves the same way Alacritty does right now. I think keeping it the way it is right now should be fine, especially because it improves discoverability.

If others give similar feedback to the one you've given me, adjusting the default behavior should be reconsidered. However until then I'd just recommend to tweak things to your hearts content.

> There doesn't seem to be a whitelist of valid schemas, so clicking on asdf:// opens firefox for me. I'm not sure if there's a way to abuse this for an attacker

asdf:// is a perfectly valid link, which is why it's recognized and opened. It would be possible to restrict that, but I did not want to limit it artificially when it's possible to support it without any problems.

I'd recommend just not clicking on things that you don't want to open (normal links could be abused too!). But #1618 should help with knowing that what you're clicking on is recognized as an URL, so that would probably help.

kpcyrd commented:
I ran into an issue with the link detection: `asdf https://example.com/, asdf` detects `https://example.com/,` as a valid url. I think trailing `,` and `.` should be ignored.

jwilm commented:
@kpcyrd thanks for the report; would you mind creating a ticket for it?

Locking discussion here to encourage users to open issues instead.
